wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -134,35 +139,122 @@ class KotlinGenerator private constructor(
             .build())
         .addModifiers(OVERRIDE)
+
         .returns(builderClassName)
-        .addStatement("return %T(this.copy())", builderClassName)
+
+    if (!javaInterOp) {
+      return funBuilder
+          .addStatement("return %T(this.copy())", builderClassName)
+          .build()
+    }
+
+    val nameAllocator = nameAllocator(type)
+
+    funBuilder.addStatement("val builder = Builder()")
+
+    type.fields().forEach { field ->
+      val fieldName = nameAllocator.get(field)
+      funBuilder.addStatement("builder.%L = %L", fieldName, fieldName)
+    }
+
+    return funBuilder
+        .addStatement("builder.addUnknownFields(unknownFields());")
+        .addStatement("return builder", builderClassName)
         .build()
   }
 
-  private fun generateBuilderClass(className: ClassName, builderClassName: ClassName): TypeSpec {
-    return TypeSpec.classBuilder("Builder")
-        .primaryConstructor(FunSpec.constructorBuilder()
-            .addParameter("message", className)
-            .build())
+  private fun generateBuilderClass(
+    type: MessageType,
+    className: ClassName,
+    builderClassName: ClassName
+  ): TypeSpec {
+    val builder = TypeSpec.classBuilder("Builder")
         .superclass(Message.Builder::class.asTypeName()
             .parameterizedBy(className, builderClassName))
-        .addProperty(PropertySpec.builder("message", className)
-            .addModifiers(PRIVATE)
-            .initializer("message")
-            .build())
-        .addFunction(FunSpec.builder("build")
-            .addModifiers(OVERRIDE)
-            .returns(className)
-            .addStatement("return message")
+
+    if (!javaInterOp) {
+      return builder
+          .primaryConstructor(FunSpec.constructorBuilder()
+              .addParameter("message", className)
+              .build())
+          .addProperty(PropertySpec.builder("message", className)
+              .addModifiers(PRIVATE)
+              .initializer("message")
+              .build())
+          .addFunction(FunSpec.builder("build")
+              .addModifiers(OVERRIDE)
+              .returns(className)
+              .addStatement("return message")
+              .build())
+          .build()
+    }
+
+    val nameAllocator = nameAllocator(type)
+    val builderClass = className.nestedClass("Builder")
+    val indentation = " ".repeat(4)
+    val internalClass = ClassName("com.squareup.wire.internal", "Internal")
+
+    val returnBody = CodeBlock.builder()
+        .add("return %T(\n", className)
+
+
+    type.fields().forEach { field ->
+      val fieldName = nameAllocator.get(field)
+      var throwExceptionBlock = CodeBlock.of("")
+
+      if (!field.isRepeated && !field.isOptional && field.default == null) {
+        throwExceptionBlock = CodeBlock.of(" ?: throw %T.%L(%L, %S)",
+            internalClass, "missingRequiredFields", field.name(), field.name())
```

Comment:
nit: move `"missingRequiredFields"` inside the template and remove the first `%L`